pkg/binding/add.go

```diff
@@ -81,3 +105,90 @@ func (o *BindingClient) AddBinding(bindingName string, bindAsFiles bool, unstruc
 
 	return libdevfile.AddKubernetesComponentToDevfile(string(yamlDesc), serviceBinding.Name, obj)
 }
+
+func (o *BindingClient) AddBinding(
+	flags map[string]string,
+	bindingName string,
+	bindAsFiles bool,
+	unstructuredService unstructured.Unstructured,
+	workloadName string,
+	workloadGVK schema.GroupVersionKind,
+) error {
+	service, err := o.kubernetesClient.NewServiceBindingServiceObject(unstructuredService, bindingName)
+	if err != nil {
+		return err
+	}
+
+	serviceBinding := kclient.NewServiceBindingObject(bindingName, bindAsFiles, workloadName, workloadGVK, []sboApi.Mapping{}, []sboApi.Service{service}, sboApi.ServiceBindingStatus{})
+
+	var backend backendpkg.AddBindingBackend
+	if len(flags) == 0 {
+		backend = o.interactiveBackend
+	} else {
+		backend = o.flagsBackend
+	}
+
+	var options []asker.CreationOption
+	for len(options) == 0 {
+		options, err = backend.SelectCreationOptions(flags)
+		if err != nil {
+			return err
+		}
+	}
+
+	// Note: we cannot directly marshal the serviceBinding object to yaml because it doesn't do that in the correct k8s manifest format
+	serviceBindingUnstructured, err := kclient.ConvertK8sResourceToUnstructured(serviceBinding)
+	if err != nil {
+		return err
+	}
+	yamlDesc, err := yaml.Marshal(serviceBindingUnstructured.UnstructuredContent())
+	if err != nil {
+		return err
+	}
+
+	var filename string
+	for _, option := range options {
+		if option == asker.OutputToFile {
+			filename, err = backend.AskOutputFilePath(flags, filepath.Join("kubernetes", serviceBinding.GetName()+".yaml"))
+			if err != nil {
+				return err
+			}
+			break
+		}
+	}
+
+	for _, option := range options {
+		switch option {
+		case asker.OutputToFile:
+			err = os.MkdirAll(filepath.Dir(filename), 0750)
+			if err != nil {
+				return err
+			}
+			err = os.WriteFile(filename, yamlDesc, 0600)
+			if err != nil {
+				return err
+			}
+
+		case asker.OutputToStdout:
+			fmt.Println(string(yamlDesc))
+
+		case asker.CreateOnCluster:
+			_, err = o.kubernetesClient.PatchDynamicResource(serviceBindingUnstructured)
+			if err != nil {
+				return err
+			}
+		}
```

Comment:
IMO, we are doing a set of things in the business layer that ought to be done by CLI, or perhaps by a different method in `kclient` package. For example, creating a file and writing to it, and printing something using `fmt.Println` should not be happening in business layer. And the call to `o.kubernetesClient.PatchDynamicResource(serviceBindingUnstructured)` should be happening from the CLI layer.

This method should only help generate a binding and not get into the business of doing anything else. In that sense, it should probably be called `CreateBindingWithoutDevfile` and return the created binding.

Probably as you said on https://github.com/redhat-developer/odo/pull/5858/files#r905903094, this was done by mistake.

Reply:
I agree for the log.Print and fmt.Print.

For the creation of file and resource, I'm not sure

Reply:
> For the creation of file and resource, I'm not sure

IMO, business layer should help do things that CLI layer absolutely can't/shouldn't. Also, if I imagine this blurb of code from a unit test perspective, it's not a good candidate because of the sheer amount of things that this function does due to the different cases it handles. 

OTOH, there are places in `pkg/binding/` where we do things based on certain backend. But if we think of the business layer as the API, it's difficult to use this function for something helpful without being tied tightly to what it does and how it does it.

@feloy maybe we should discuss `pkg/binding/` in a separate issue and refactor it at some point?

Reply:
> > For the creation of file and resource, I'm not sure
> 
> IMO, business layer should help do things that CLI layer absolutely can't/shouldn't. Also, if I imagine this blurb of code from a unit test perspective, it's not a good candidate because of the sheer amount of things that this function does due to the different cases it handles.
> 

For me, the business layer does what is common to all frontends (CLI, plugin, HTML interface?, etc). 

To get the plugin as an example, it would use the flags (and so the flag backend) and the JSON output, and in this case, the file would be created the same way, and the resource would be created in the cluster the same way. The only difference would be how the result is displayed. For JSON output, it would be displayed in the JSON, not directly to stdout as raw text.